src/UsgsAstroPlugin.cpp

```diff
@@ -0,0 +1,306 @@
+  size_t lastIndex = imageFilename.find_last_of(".");
+  std::string baseName = imageFilename.substr(0, lastIndex);
+  std::string isdFilename = baseName.append(".json");
+
+  try {
+    std::ifstream isd_sidecar(isdFilename);
+    json jsonisd;
+    isd_sidecar >> jsonisd;
+    return jsonisd.dump();
+
+  } catch (...) {
+    std::string errorMessage = "Could not read metadata file associated with image: ";
+    errorMessage.append(isdFilename);
+    throw csm::Error(csm::Error::FILE_READ, errorMessage,
+                     "UsgsAstroPlugin::loadImageSupportData");
+  }
+}
+
+
+std::string UsgsAstroPlugin::getModelNameFromModelState(const std::string &modelState,
+                                                   csm::WarningList *warnings) const {
+  auto state = json::parse(modelState);
+
+  std::string name = state.value<std::string>("model_name", "");
+
+  if (name == "") {
+      csm::Error::ErrorType aErrorType = csm::Error::INVALID_SENSOR_MODEL_STATE;
+      std::string aMessage = "No 'model_name' key in the model state object.";
+      std::string aFunction = "UsgsAstroPlugin::getModelNameFromModelState";
+      csm::Error csmErr(aErrorType, aMessage, aFunction);
+      throw(csmErr);
+  }
+
+  // if (MODEL_KEYWORDS.find(name) == MODEL_KEYWORDS.end()){
+  //     csm::Error::ErrorType aErrorType = csm::Error::SENSOR_MODEL_NOT_SUPPORTED;
+  //     std::string aMessage = "Sensor model not supported.";
+  //     std::string aFunction = "UsgsAstroPlugin::getModelNameFromModelState()";
+  //     csm::Error csmErr(aErrorType, aMessage, aFunction);
+  //     throw(csmErr);
+  // }
+
+  return name;
+}
+
+
+bool UsgsAstroPlugin::canISDBeConvertedToModelState(const csm::Isd &imageSupportData,
+                                               const std::string &modelName,
+                                               csm::WarningList *warnings) const {
+  try {
+       convertISDToModelState(imageSupportData, modelName);
+  }
+  catch(...) {
+      // key didn't exist
+      return false;
+  }
+  return true;
+}
+
+
+std::string UsgsAstroPlugin::getStateFromISD(csm::Isd imageSupportData) const {
+    std::string stringIsd = loadImageSupportData(imageSupportData);
+    json jsonIsd = json::parse(stringIsd);
+    return convertISDToModelState(imageSupportData, jsonIsd.at("modelName"));
+}
+
+
+std::string UsgsAstroPlugin::convertISDToModelState(const csm::Isd &imageSupportData,
+                                               const std::string &modelName,
+                                               csm::WarningList *warnings) const {
+
+  csm::Model* sensor_model = constructModelFromISD(
+                             imageSupportData, modelName);
+
+  if (sensor_model == 0) {
+      csm::Error::ErrorType aErrorType = csm::Error::ISD_NOT_SUPPORTED;
+      std::string aMessage = "ISD not supported: ";
+      std::string aFunction = "UsgsAstroPlugin::convertISDToModelState()";
+      throw csm::Error(aErrorType, aMessage, aFunction);
+  }
```

Comment:
Chances are if it's not supported, constructModelFromISD is going to throw an exception. I don't know a case in which it would return NULL without an exception.